### PR TITLE
Fixed #306 deadlock between RTPTranslatorImpl and MediaStreamImpl tha…

### LINKED ISSUE #306: Deadlock between RTPTranslatorImpl and MediaStreamImpl
There appears to be a lock order inversion that causes a circular dependency between RTPTranslatorImpl and MediaStreamImpl.  This failure mode is fatal and we need to restart the bridge when it happens.

I like to understand the code a bit more before we attempt a fix.  Observe that the synchronization block of dynamicRTPPayloadTypes is very broad.  Could we clarify the intent of the dynamicRTPPayloadTypes lock?  A more fine grained locking would fix this deadlock.

MediaStreamImpl.addDynamicRTPPayloadType
https://github.com/jitsi/libjitsi/blob/master/src/org/jitsi/impl/neomedia/MediaStreamImpl.java#L449

## Deadlocked threads:

There is an inversion in locking order which results in the deadlock.

**Thread 1**
RTPTranslatorImpl.didRead acquires read lock
https://github.com/jitsi/libjitsi/blob/master/src/org/jitsi/impl/neomedia/rtp/translator/RTPTranslatorImpl.java#L594
RTCPFeedbackMessageSender. maybeStopRequesting
https://github.com/jitsi/libjitsi/blob/master/src/org/jitsi/impl/neomedia/rtp/translator/RTCPFeedbackMessageSender.java#L288
MediaStreamImpl. isKeyFrame
https://github.com/jitsi/libjitsi/blob/master/src/org/jitsi/impl/neomedia/MediaStreamImpl.java#L3915
MediaStreamImpl. getPayloadBlock
https://github.com/jitsi/libjitsi/blob/master/src/org/jitsi/impl/neomedia/MediaStreamImpl.java#L3994
MediaStreamImpl. getDynamicRTPPayloadType acquires dynamicRTPPayloadTypes
https://github.com/jitsi/libjitsi/blob/master/src/org/jitsi/impl/neomedia/MediaStreamImpl.java#L1498

**Thread 2**
MediaStreamImpl.addDynamicRTPPayloadType acquires dynamicRTPPayloadTypes
https://github.com/jitsi/libjitsi/blob/master/src/org/jitsi/impl/neomedia/MediaStreamImpl.java#L482
StreamRTPManager. addFormat
https://github.com/jitsi/libjitsi/blob/master/src/org/jitsi/impl/neomedia/rtp/StreamRTPManager.java#L81
RTPTranslatorImpl. addFormat acquires write lock
https://github.com/jitsi/libjitsi/blob/master/src/org/jitsi/impl/neomedia/rtp/translator/RTPTranslatorImpl.java#L293

<hr>
A stacktrack from our deployed bridge on libjitsi version 891-0-gafa086b3.  We also believe this also occurs in the head of libjitsi.

```"org.jitsi.impl.neomedia.rtp.translator.PushSourceStreamImpl" #24130 daemon prio=5 os_prio=0 tid=0x00007f74f44a2000 nid=0x339f waiting for monitor entry [0x00007f73f5dd9000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.jitsi.impl.neomedia.MediaStreamImpl.getDynamicRTPPayloadTypes(MediaStreamImpl.java:1503)
	- waiting to lock <0x00000006d875f710> (a java.util.HashMap)  dynamicRTPPayloadTypes
	at org.jitsi.impl.neomedia.MediaStreamImpl.isKeyFrame(MediaStreamImpl.java:3704)
	at org.jitsi.impl.neomedia.rtp.translator.RTCPFeedbackMessageSender$KeyframeRequester.maybeStopRequesting(RTCPFeedbackMessageSender.java:270)
	at org.jitsi.impl.neomedia.rtp.translator.RTCPFeedbackMessageSender.maybeStopRequesting(RTCPFeedbackMessageSender.java:196)
	- locked ReentrantReadWriteLock$ReadLock lock <0x00000006d644c988>
	at org.jitsi.impl.neomedia.rtp.translator.RTPTranslatorImpl.didRead(RTPTranslatorImpl.java:594)
	at org.jitsi.impl.neomedia.rtp.translator.PushSourceStreamImpl.read(PushSourceStreamImpl.java:271)
	at net.sf.fmj.media.rtp.util.RTPPacketReceiver.transferData(RTPPacketReceiver.java:168)
	at org.jitsi.impl.neomedia.rtp.translator.PushSourceStreamImpl.run(PushSourceStreamImpl.java:328)
	at java.lang.Thread.run(Thread.java:745)


   "qtp1481451788-26965" #26965 prio=5 os_prio=0 tid=0x00007f74f5356800 nid=0x3b3e waiting on condition [0x00007f73a24a7000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006d644c988> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:870)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1199)
	at java.util.concurrent.locks.ReentrantReadWriteLock$WriteLock.lock(ReentrantReadWriteLock.java:943)
	at org.jitsi.impl.neomedia.rtp.translator.RTPTranslatorImpl.addFormat(RTPTranslatorImpl.java:293)
	at org.jitsi.impl.neomedia.rtp.StreamRTPManager.addFormat(StreamRTPManager.java:118)
	at org.jitsi.impl.neomedia.MediaStreamImpl.addDynamicRTPPayloadType(MediaStreamImpl.java:494)
	- locked <0x00000006d875f710> (a java.util.HashMap)   dynamicRTPPayloadTypes
	at org.jitsi.videobridge.RtpChannel.setPayloadTypes(RtpChannel.java:1363)
	at org.jitsi.videobridge.VideoChannel.setPayloadTypes(VideoChannel.java:867)
	at org.jitsi.videobridge.Videobridge.handleColibriConferenceIQ(Videobridge.java:942)
   ...```

### PR DESCRIPTION
**Pull request by avila-devlogic**
…t causes JVB to leak conferences

### CODE REVIEW

**Comment by bgrozev** [comment]
Jenkins test this please

**Comment by avila-devlogic** [comment]
@bgrozev , could someone take a look at my change ?

Thanks

**Comment by bgrozev** [comment]
It's on my (personal) list. I'll try and take a look next week.

**Comment by bgrozev** [comment]
Thank you for the contribution! We can get it merged, as soon as we find your CLA -- have you already signed one?

File: src/org/jitsi/impl/neomedia/MediaStreamImpl.java [hunk 1]
```diff
@@ -1495,19 +1495,16 @@ public MediaDirection getDirection()
      */
     public byte getDynamicRTPPayloadType(String encoding)
     {
-        synchronized (dynamicRTPPayloadTypes)
+        for (Map.Entry<Byte, MediaFormat> dynamicRTPPayloadType
```
Comment (gpolitis) [inline src/org/jitsi/impl/neomedia/MediaStreamImpl.java]: This is not thread safe, so we can't remove the synchronized block.

Comment (bgrozev) [inline src/org/jitsi/impl/neomedia/MediaStreamImpl.java]: What do you mean exactly? The getter returns a copy.

**Comment by gpolitis** [comment]
There are thread safety issues with this code, so I don't think we're ready to merge it. What is the exact issue that you're attempting to fix?

**Comment by bgrozev** [comment]
> There are thread safety issues with this code, so I don't think we're ready to merge it. What is the exact issue that you're attempting to fix?

The deadlock reported in #306

File: src/org/jitsi/impl/neomedia/MediaStreamImpl.java (hunk 1 above)
Comment (gpolitis) [inline src/org/jitsi/impl/neomedia/MediaStreamImpl.java]: No, not really, it returns a view of the mappings contained in this map.

Comment (gpolitis) [inline src/org/jitsi/impl/neomedia/MediaStreamImpl.java]: I'm sorry, I misread the code. There's an intermediate call that actually copies the set.

**Comment by avila-devlogic** [comment]
@bgrozev , @gpolitis thanks for the review.

@bgrozev, I'm contributing under the Highfive's corporate CLA.
@imedwei might clarify more if needed.

**Pull request merged**